### PR TITLE
[ZMQ] Notify when an IS double spend is attempted

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
Not tested

### CODE REVIEW

File: src/instantx.cpp [hunk 1]
```diff
@@ -117,7 +117,8 @@ bool CInstantSend::ProcessTxLockRequest(const CTxLockRequest& txLockRequest, CCo
                 if(hash != txLockRequest.GetHash()) {
                     LogPrint("instantsend", "CInstantSend::ProcessTxLockRequest -- Double spend attempt! %s\n", txin.prevout.ToStringShort());
                     // do not fail here, let it go and see which one will get the votes to be locked
-                    // TODO: notify zmq+script
+                    // NOTIFY SCRIPTS + ZMQ
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/instantx.cpp]: I believe this has been the format for zmq notification comments. I was under the assumption notifying scripts is the same thing as zmq

File: src/instantx.cpp [hunk 2]
```diff
@@ -117,7 +117,8 @@ bool CInstantSend::ProcessTxLockRequest(const CTxLockRequest& txLockRequest, CCo
...
+                    GetMainSignals().NotifyInstantSendDoubleSpendAttempt(txin.prevout);
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/instantx.cpp]: So you're thinking `Notify(CTransaction ours, CTransaction conflicts With Ours) ` type of thing? I think that would probably be good

**Comment by PastaPastaPasta** [comment]
@UdjinM6 That looks pretty good to me, thoughts?

File: src/instantx.cpp [hunk 3]
```diff
@@ -117,8 +117,8 @@ bool CInstantSend::ProcessTxLockRequest(const CTxLockRequest& txLockRequest, CCo
                 if(hash != txLockRequest.GetHash()) {
                     LogPrint("instantsend", "CInstantSend::ProcessTxLockRequest -- Double spend attempt! %s\n", txin.prevout.ToStringShort());
                     // do not fail here, let it go and see which one will get the votes to be locked
-                    // NOTIFY SCRIPTS + ZMQ
-                    GetMainSignals().NotifyInstantSendDoubleSpendAttempt(txin.prevout);
+                    // NOTIFY ZMQ
+                    GetMainSignals().NotifyInstantSendDoubleSpendAttempt(txHash, hash);
```
Comment (UdjinM6) [inline src/instantx.cpp]: Should pass txes here, not hashes, just like you described in your previous comment re the issue https://github.com/dashpay/dash/pull/2262#discussion_r215264029. There is no use of "raw" hashes, the whole point of "raw" messages is to provide the _object_ that triggered the notification.

File: src/init.cpp [hunk 4]
```diff
@@ -508,11 +508,11 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-zmqpubhashtxlock=<address>", _("Enable publish hash transaction (locked via InstantSend) in <address>"));
     strUsage += HelpMessageOpt("-zmqpubhashgovernancevote=<address>", _("Enable publish hash of governance votes in <address>"));
     strUsage += HelpMessageOpt("-zmqpubhashgovernanceobject=<address>", _("Enable publish hash of governance objects (like proposals) in <address>"));
-    strUsage += HelpMessageOpt("-zmqpubhashinstantsenddoublespend=<address>", _("Enable publish of attempted InstantSend double sends in <address>"));
+    strUsage += HelpMessageOpt("-zmqpubhashinstantsenddoublespend=<address>", _("Enable hash publish of attempted InstantSend double spend in <address>"));
```
Comment (UdjinM6) [inline src/init.cpp]: s/hash publish/publish transaction hashes/

File: src/init.cpp [hunk 5]
```diff
@@ -508,11 +508,11 @@ std::string HelpMessage(HelpMessageMode mode)
...
     strUsage += HelpMessageOpt("-zmqpubrawblock=<address>", _("Enable publish raw block in <address>"));
     strUsage += HelpMessageOpt("-zmqpubrawtx=<address>", _("Enable publish raw transaction in <address>"));
     strUsage += HelpMessageOpt("-zmqpubrawtxlock=<address>", _("Enable publish raw transaction (locked via InstantSend) in <address>"));
-    strUsage += HelpMessageOpt("-zmqpubrawinstantsenddoublespend=<address>", _("Enable publish of attempted InstantSend double sends in <address>"));
+    strUsage += HelpMessageOpt("-zmqpubrawinstantsenddoublespend=<address>", _("Enable raw publish of attempted InstantSend double spend in <address>"));
```
Comment (UdjinM6) [inline src/init.cpp]: s/raw publish/publish raw transactions/

**Comment by UdjinM6** [comment]
Also (just remembered), you need to update `doc/zmq.md`.

File: src/instantx.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/instantx.cpp]: I tried to do such, but I could not find a way to get the CTransaction objects based on what is there

**Comment by UdjinM6** [comment]
Needs a couple of fixes, pls see https://github.com/UdjinM6/dash/commit/617f1e5e8ababeeb194fb3bbfb00d00b9e990855

**Comment by PastaPastaPasta** [comment]
Ahh, that does look a lot cleaner.

**Pull request merged**